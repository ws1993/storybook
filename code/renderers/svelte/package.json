{
  "name": "@storybook/svelte",
  "version": "9.0.0-alpha.0",
  "description": "Storybook Svelte renderer",
  "keywords": [
    "storybook"
  ],
  "homepage": "https://github.com/storybookjs/storybook/tree/next/code/renderers/svelte",
  "bugs": {
    "url": "https://github.com/storybookjs/storybook/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/storybookjs/storybook.git",
    "directory": "code/renderers/svelte"
  },
  "funding": {
    "type": "opencollective",
    "url": "https://opencollective.com/storybook"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    },
    "./experimental-playwright": {
      "types": "./dist/playwright.d.ts",
      "import": "./dist/playwright.mjs",
      "require": "./dist/playwright.js"
    },
    "./preset": "./preset.js",
    "./dist/entry-preview.mjs": "./dist/entry-preview.mjs",
    "./dist/entry-preview-docs.mjs": "./dist/entry-preview-docs.mjs",
    "./package.json": "./package.json",
    "./internal/PreviewRender.svelte": "./dist/components/PreviewRender.svelte",
    "./internal/SlotDecorator.svelte": "./dist/components/SlotDecorator.svelte",
    "./internal/AddStorybookIdDecorator.svelte": "./dist/components/AddStorybookIdDecorator.svelte",
    "./internal/createSvelte5Props": "./dist/createSvelte5Props.svelte.js"
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist/**/*",
    "template/cli/**/*",
    "README.md",
    "*.js",
    "*.d.ts",
    "!src/**/*"
  ],
  "scripts": {
    "check": "svelte-check",
    "prep": "jiti ../../../scripts/prepare/bundle.ts"
  },
  "dependencies": {
    "@storybook/csf": "0.1.12",
    "@storybook/global": "^5.0.0",
    "sveltedoc-parser": "^4.2.1",
    "ts-dedent": "^2.0.0",
    "type-fest": "~2.19"
  },
  "devDependencies": {
    "@sveltejs/vite-plugin-svelte": "^5.0.3",
    "@testing-library/svelte": "^5.2.4",
    "expect-type": "^1.1.0",
    "svelte": "^5.19.2",
    "svelte-check": "^4.1.4",
    "typescript": "^5.7.3",
    "vite": "^6.0.11"
  },
  "peerDependencies": {
    "storybook": "workspace:^",
    "svelte": "^4.0.0 || ^5.0.0"
  },
  "engines": {
    "node": ">=18.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "bundler": {
    "post": "./scripts/copy-unbundled-to-dist.ts",
    "entries": [
      "./src/index.ts",
      "./src/preset.ts",
      "./src/entry-preview.ts",
      "./src/entry-preview-docs.ts",
      "./src/playwright.ts"
    ],
    "platform": "browser"
  },
  "gitHead": "e6a7fd8a655c69780bc20b9749c2699e44beae16"
}
