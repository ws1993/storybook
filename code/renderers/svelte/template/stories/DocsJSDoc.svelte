<script>
  /**
   * @typedef {100 | 1000 | 10000} LiteralNumbers
   */

  /**
   * @typedef {"apple" | "grape" | "orange"} LiteralStrings
   */

  /**
   * @typedef {Object} MyObject
   * @property {string} foo
   * @property {number} bar
   */

  /**
   * @type {boolean} Boolean
   */
  export let boolean = true;
  /**
   * @type {string} String
   */
  export let string = 'default';
  /**
   * @type {string} String (required)
   */
  export let stringRequired;
  /**
   * @type {number} Number
   */
  export let number = 123;
  /**
   * True literal
   * @type {true=}
   */
  export let trueLiteral = undefined;
  /**
   * Null
   * @type {null}
   */
  export let nullValue = null;
  /**
   * Undefined
   * @type {undefined}
   */
  export let undefinedValue = undefined;
  /**
   * Any
   * @type {any}
   */
  export let any = null;
  /**
   * Array of numbers
   * @type {number[]}
   */
  export let arrayOfNumbers = [1, 20, 300];
  /**
   * Union of literal strings
   * @type {LiteralStrings=}
   */
  export let unionLiteralStrings = 'apple';
  /**
   * Union of literal numbers
   * @type {LiteralNumbers=}
   */
  export let unionLiteralNumbers = 1000;
  /**
   * @type {MyObject=} Object
   */
  export let object = undefined;
  /**
   * Inline object
   * @type {{ foo: string; bar: number } =}
   */
  export let inlineObject = undefined;
  /**
   * Record
   * @type {Record<string, number>}
   */
  export let record = { a: 1, b: 2 };
  /**
   * Date
   * @type {Date}
   */
  export let date = new Date('20 Jan 1983');
  /**
   * Union of types
   * @type {number | string =}
   */
  export let unionTypes = 123;
  /**
   * Intersection of types
   * @type {{ a: number } & { b: string } =}
   */
  export let intersection = undefined;
  /**
   * Event callback function
   * @type {(event: MouseEvent) => number}
   */
  export let func = () => 10;
</script>

<div>Docs: JSDoc</div>
