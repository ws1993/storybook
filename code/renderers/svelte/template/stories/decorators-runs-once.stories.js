import ButtonJavaScript from './views/ButtonJavaScript.svelte';

export default {
  component: ButtonJavaScript,
  args: {
    primary: true,
  },
  decorators: [
    (Story) => {
      console.log('decorator called');
      return Story();
    },
  ],
};

export const Default = {};
