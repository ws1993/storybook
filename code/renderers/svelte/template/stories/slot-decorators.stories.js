import BorderDecoratorBlue from './views/BorderDecoratorBlue.svelte';
import BorderDecoratorProps from './views/BorderDecoratorProps.svelte';
import BorderDecoratorRed from './views/BorderDecoratorRed.svelte';
import ButtonView from './views/ButtonJavaScript.svelte';

export default {
  component: ButtonView,
  decorators: [() => BorderDecoratorRed],
  args: {
    primary: true,
  },
  tags: ['autodocs'],
};

export const WithDefaultRedBorder = {};
export const WithBareBlueBorder = {
  decorators: [() => BorderDecoratorBlue],
};
export const WithPreparedBlueBorder = {
  decorators: [
    () => ({
      Component: BorderDecoratorBlue,
    }),
  ],
};
export const WithPropsBasedBorder = {
  decorators: [
    () => ({
      Component: BorderDecoratorProps,
      props: { color: 'green' },
    }),
  ],
};
export const WithArgsBasedBorderUnset = {
  argTypes: {
    color: { control: 'color' },
  },
  decorators: [(_, { args }) => ({ Component: BorderDecoratorProps, props: args })],
};
export const WithArgsBasedBorder = {
  ...WithArgsBasedBorderUnset,
  args: { color: 'lightblue' },
};
