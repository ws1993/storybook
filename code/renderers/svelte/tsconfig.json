{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "module": "ESNext",
    "verbatimModuleSyntax": true
  },
  "include": ["src/**/*", "src/**/*.svelte", "template/**/*", "template/**/*.svelte"]
}
