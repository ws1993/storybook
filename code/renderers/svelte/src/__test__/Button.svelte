<script lang="ts">
  import { createEventDispatcher } from 'svelte';

  export let disabled: boolean;
  export let label: string;

  const dispatch = createEventDispatcher();
</script>

<button on:click={(e) => dispatch('clicked', e)} on:dblclick on:mousemove {disabled}>
  {label}
</button>
