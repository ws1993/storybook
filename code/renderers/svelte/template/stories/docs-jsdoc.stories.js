import DocsJSDoc from './DocsJSDoc.svelte';

export default {
  component: DocsJSDoc,
  args: {},
  tags: ['autodocs'],
};

export const Primary = {};
