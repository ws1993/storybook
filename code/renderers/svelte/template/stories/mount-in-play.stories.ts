import type { StoryObj } from '@storybook/svelte';

import Button from './Button.svelte';

export default {
  component: Button,
};

export const Basic: StoryObj = {
  args: {
    disabled: true,
  },
  async play({ mount, args }) {
    await mount(Button, { props: { ...args, label: 'set in play' } });
  },
};
