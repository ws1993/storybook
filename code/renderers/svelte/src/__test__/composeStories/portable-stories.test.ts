// @vitest-environment happy-dom
/// <reference types="@testing-library/jest-dom" />;
import { cleanup, render, screen } from '@testing-library/svelte';
import { afterEach, describe, expect, it, vi } from 'vitest';

// import '@testing-library/svelte/vitest';
import { expectTypeOf } from 'expect-type';

import type { Meta } from '../..';
import { composeStories, composeStory, setProjectAnnotations } from '../../portable-stories';
import * as stories from './Button.stories';
import type Button from './Button.svelte';

setProjectAnnotations([]);

// example with composeStories, returns an object with all stories composed with args/decorators
const { CSF3Primary, LoaderStory } = composeStories(stories);

afterEach(() => {
  cleanup();
});

// example with composeStory, returns a single story composed with args/decorators
const Secondary = composeStory(stories.CSF2Secondary, stories.default);
describe('renders', () => {
  it('renders primary button with custom props via composeStory', () => {
    // We unfortunately can't do the following:
    // render(CSF3Primary.Component, { ...CSF3Primary.props, label: 'Hello world' });
    // Because the props will be passed to the first decorator of the story instead
    // of the actual component of the story. This is because of our current PreviewRender structure

    const Composed = composeStory(
      {
        ...stories.CSF3Primary,
        args: { ...stories.CSF3Primary.args, label: 'Hello world' },
      },
      stories.default
    );

    render(Composed.Component, Composed.props);
    const buttonElement = screen.getByText(/Hello world/i);
    expect(buttonElement).not.toBeNull();
  });

  it('reuses args from composed story', () => {
    render(Secondary.Component, Secondary.props);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement.textContent).toMatch(Secondary.args.label);
  });

  // TODO TypeError: component.$on is not a function - Potentially only works in Svelte 4
  it.skip('onclick handler is called', async () => {
    const onClickSpy = vi.fn();
    const { component } = render(Secondary.Component, { ...Secondary.props, onClick: onClickSpy });
    component.$on('click', onClickSpy);
    const buttonElement = screen.getByRole('button');
    buttonElement.click();
    expect(onClickSpy).toHaveBeenCalled();
  });

  it('reuses args from composeStories', () => {
    const { getByText } = render(CSF3Primary.Component, CSF3Primary.props);
    const buttonElement = getByText(/foo/i);
    expect(buttonElement).not.toBeNull();
  });

  it('should call and compose loaders data', async () => {
    await LoaderStory.load();
    const { getByTestId } = render(LoaderStory.Component, LoaderStory.props);
    expect(getByTestId('spy-data').textContent).toEqual('mockFn return value');
    expect(getByTestId('loaded-data').textContent).toEqual('loaded data');
    // spy assertions happen in the play function and should work
    await LoaderStory.run!();
  });
});

describe('projectAnnotations', () => {
  it('renders with default projectAnnotations', () => {
    setProjectAnnotations([
      {
        parameters: { injected: true },
        globalTypes: {
          locale: { defaultValue: 'en' },
        },
      },
    ]);
    const WithEnglishText = composeStory(stories.CSF2StoryWithLocale, stories.default);
    const { getByText } = render(WithEnglishText.Component, WithEnglishText.props);
    const buttonElement = getByText('Hello!');
    expect(buttonElement).not.toBeNull();
    expect(WithEnglishText.parameters?.injected).toBe(true);
  });

  it('renders with custom projectAnnotations via composeStory params', () => {
    const WithPortugueseText = composeStory(stories.CSF2StoryWithLocale, stories.default, {
      initialGlobals: { locale: 'pt' },
    });
    const { getByText } = render(WithPortugueseText.Component, WithPortugueseText.props);
    const buttonElement = getByText('OlÃ¡!');
    expect(buttonElement).not.toBeNull();
  });
});

describe('CSF3', () => {
  it('renders with inferred globalRender', () => {
    const Primary = composeStory(stories.CSF3Button, stories.default);
    render(Primary.Component, Primary.props);
    const buttonElement = screen.getByText(/foo/i);
    expect(buttonElement).not.toBeNull();
  });

  it('renders with custom render function', () => {
    const Primary = composeStory(stories.CSF3ButtonWithRender, stories.default);

    render(Primary.Component, Primary.props);
    expect(screen.getByTestId('custom-render')).not.toBeNull();
  });

  it('renders with play function without canvas element', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    await CSF3InputFieldFilled.run();

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });

  it('renders with play function with canvas element', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    const div = document.createElement('div');
    document.body.appendChild(div);

    await CSF3InputFieldFilled.run({ canvasElement: div });

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');

    document.body.removeChild(div);
  });
});

describe('ComposeStories types', () => {
  // this file tests Typescript types that's why there are no assertions
  it('Should support typescript operators', () => {
    type ComposeStoriesParam = Parameters<typeof composeStories>[0];

    expectTypeOf({
      ...stories,
      default: stories.default as Meta<typeof Button>,
    }).toMatchTypeOf<ComposeStoriesParam>();

    expectTypeOf({
      ...stories,
      // TODO: @JReinhold maybe you can figure this error out?
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore-error (Type 'IsomorphicComponent<PropsWithChildren<{ primary?: boolean | undefined; backgroundColor?: string | undefined; size?: "small" | "medium" | "large" | undefined; label?: string | undefined; }, { default: {}; }>, { ...; } & { ...; }, { ...; }, {}, string>' has no properties in common with type 'PropsWithChildren<{ primary?: boolean | undefined; backgroundColor?: string | undefined; size?: "small" | "medium" | "large" | undefined; label?: string | undefined; }, { default: {}; }>'. )
      default: stories.default satisfies Meta<typeof Button>,
    }).toMatchTypeOf<ComposeStoriesParam>();
  });
});

// Batch snapshot testing
const testCases = Object.values(composeStories(stories)).map(
  (Story) => [Story.storyName, Story] as [string, typeof Story]
);
it.each(testCases)('Renders %s story', async (_storyName, Story) => {
  if (_storyName === 'CSF2StoryWithLocale') {
    return;
  }
  await Story.run();
  expect(document.body).toMatchSnapshot();
});
