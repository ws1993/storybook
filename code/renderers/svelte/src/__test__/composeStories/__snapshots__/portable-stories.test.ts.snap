// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders CSF2Secondary story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      style=""
      type="button"
    >
      label coming from story args! 
      <!---->
    </button>
    
    
    
  </div>
</body>
`;

exports[`Renders CSF2StoryWithParamsAndDecorator story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      data-testid="local-decorator"
      style="margin: 3em;"
    >
      <!---->
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        style=""
        type="button"
      >
        foo 
        <!---->
      </button>
      
      <!---->
    </div>
    
    
    
    
    
  </div>
</body>
`;

exports[`Renders CSF3Button story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      style=""
      type="button"
    >
      foo 
      <!---->
    </button>
    
    
    
  </div>
</body>
`;

exports[`Renders CSF3ButtonWithRender story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <div>
      <p
        data-testid="custom-render"
      >
        I am a custom render function
      </p>
       
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        style=""
        type="button"
      >
        foo 
        <!---->
      </button>
      <!---->
    </div>
    
    
    
  </div>
</body>
`;

exports[`Renders CSF3InputFieldFilled story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <input
      data-testid="input"
      formaction="http://localhost:3000/"
      formmethod=""
    />
    
    
    
  </div>
</body>
`;

exports[`Renders CSF3Primary story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <button
      class="storybook-button storybook-button--large storybook-button--primary"
      style=""
      type="button"
    >
      foo 
      <!---->
    </button>
    
    
    
  </div>
</body>
`;

exports[`Renders LoaderStory story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <div>
      <div
        data-testid="loaded-data"
      >
        loaded data
      </div>
       
      <div
        data-testid="spy-data"
      >
        mockFn return value
      </div>
    </div>
    
    
    
  </div>
</body>
`;

exports[`Renders NewStory story 1`] = `
<body>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      data-testid="local-decorator"
      style="margin: 3em;"
    >
      <!---->
      <button
        class="storybook-button storybook-button--large storybook-button--primary"
        style=""
        type="button"
      >
        foo 
        <!---->
      </button>
      
      <!---->
    </div>
    
    
    
    
    
  </div>
</body>
`;
